docs/docs/understanding/workflows/extending.md

```diff
@@ -0,0 +1,161 @@
+# Extending workflows
+
+Another great feature of workflows is their extensibility. You can take workflows written by others or built-ins from LlamaIndex and extend them to customize them to your needs. Let's look at two ways to do that.
+
+## Subclassing workflows
+
+Workflows are just regular Python classes, which means you can subclass them to add new functionality. For example, let's say you have an agentic workflow that does some processing and then sends an email. You can subclass the workflow to add an extra step to send a text message as well.
+
+Here's our base workflow:
+
+```python
+from llama_index.core.workflow import (
+    StartEvent,
+    StopEvent,
+    Workflow,
+    step,
+    Event,
+    Context,
+)
+
+
+class Step2Event(Event):
+    query: str
+
+
+class Step3Event(Event):
+    query: str
+
+
+class MainWorkflow(Workflow):
+    @step()
+    async def start(self, ev: StartEvent) -> Step2Event:
+        print("Starting up")
+        return Step2Event(query=ev.query)
+
+    @step()
+    async def step_two(self, ev: Step2Event) -> Step3Event:
+        print("Sending an email")
+        return Step3Event(query=ev.query)
+
+    @step()
+    async def step_three(self, ev: Step3Event) -> StopEvent:
+        print("Finishing up")
+        return StopEvent(result=ev.query)
+```
+
+If we run this:
+
+```python
+w = MainWorkflow(timeout=10, verbose=False)
+result = await w.run(query="Initial query")
+print(result)
+```
+
+We get:
+
+```
+Starting up
+Sending an email
+Finishing up
+Initial query
+```
+
+Now let's subclass this workflow to send a text message as well:
+
+```python
+class Step2BEvent(Event):
+    query: str
+
+
+class CustomWorkflow(MainWorkflow):
+    @step()
+    async def step_two(self, ev: Step2Event) -> Step2BEvent:
+        print("Sending an email")
+        return Step2BEvent(query=ev.query)
+
+    @step()
+    async def step_two_b(self, ev: Step2BEvent) -> Step3Event:
+        print("Also sending a text message")
+        return Step3Event(query=ev.query)
+```
+
+Which will instead give us
+
+```
+Starting up
+Sending an email
+Also sending a text message
+Finishing up
+Initial query
+```
+
+We can visualize the subclassed workflow and it will show all the steps, like this:
+
+```python
+draw_all_possible_flows(CustomWorkflow, "custom_workflow.html")
+```
+
+![Custom workflow](subclass.png)
+
+## Adding service flows
```

Comment:
imo the more intuitive concept here is "Hierarchical Workflows" - for instance, you might compose smaller, reusable workflows into larger workflows that operate across systems

Reply:
Not sure what you're asking here. Are you saying we should rename Service Workflows to Hierarchical Workflows?

Reply:
yeah basically (or nested workflows). it feels a bit more intuitive to me? lmk if you disagree @logan-markewich @masci

Reply:
I agree, I would think service workflows would involve inter-service communication (e.g. API calls).

Reply:
I was calling the sub-flows, but then the code calls them services so I was sticking to that.

Reply:
yea the code calls them services, I've been calling them "nested workflows" -- not sure if we should change the code or just the docs 😅

Reply:
nested workflows works too (i like it the same as hierarchical workflows)

Reply:
> I would think service workflows would involve inter-service communication (e.g. API calls).

That's the very reason why we went with "services": once deployed, these workflows will communicate with each other by api calls. Another reason why we went with "services" and not "workflows" is to be able in the future to  wrap in a service more APIs than just workflow instances - think of llama-agents providing a built-in caching service that's not a workflow (one piece we're missing to get there: a `Service` protocol or interface, to let steps call services not caring whether they're workflows or something else).

I see this more as a "flat" relationship, with one main workflow (the basic unit of work) referencing others. In this scenario, the hierarchical (or nested) nature of one workflow that's referenced in turn is actually hidden. Under this assumption, I think the docs go in the right direction, explaining how one workflow can leverage others by referencing them as services, avoiding code duplication and encouraging a modular design of AI applications.

If we want to make hierarchies more explicit in the design, I think we should change the code and remove the "service" abstraction altogether to avoid confusion. For example, the code in the docs would become `w.add_subworkflow(reflection_workflow=CustomSubflow())`, the `ServiceManager` would become `SubWorkflowsManager` and so on and so forth. Note that if "everything is a workflow", runing in llama-agents would be less flexible.